### PR TITLE
nrf,sam,rp2040: add machine.DeviceID function

### PR DESCRIPTION
**Pull request by kenbell**
This seems like a common feature and is super useful in network/lora/mesh scenarios to separate devices without flash config.

I've implemented on 3 chipsets nrf,sam,rp2040 so I'm confident we can have a common API.

### CODE REVIEW

File: src/machine/machine_atsam.go [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+// flash chip at time of manufacture.
+//
+// It's possible that two different vendors may allocate
+// the same HardwareID, so callers should take this into
+// account if needing to generate a globally unique id.
+//
+// The length of the hardware ID is vendor-specific, but
+// 8 bytes (64 bits) and 16 bytes (128 bits) are common.
+func HardwareID() []byte {
+	id := make([]byte, 16)
```
Comment (deadprogram) [inline src/machine/machine_atsam.go]: Perhaps avoid the heap allocation here?

```
var hardwareID [16]byte

func HardwareID() []byte {
    ...
    return hardwareID[:]
}
```

Comment (kenbell) [reply src/machine/machine_atsam.go]: thanks for reviewing.  yeah, i like that change - will update the PR

File: src/machine/machine_rp2040_flash.go [hunk 2]
```diff
@@ -13,6 +13,33 @@ func EnterBootloader() {
...
+//
+// It's possible that two different vendors may allocate
+// the same HardwareID, so callers should take this into
+// account if needing to generate a globally unique id.
+//
+// The length of the hardware ID is vendor-specific, but
+// 8 bytes (64 bits) is common.
+func HardwareID() []byte {
+	// 13 = 1 + FLASH_RUID_DUMMY_BYTES(4) + FLASH_RUID_DATA_BYTES(8)
+	txBuf := make([]byte, 13)
```
Comment (deadprogram) [inline src/machine/machine_rp2040_flash.go]: Should also do something about these heap allocations.

File: src/machine/machine_rp2040_rom.go [hunk 3]
```diff
@@ -132,13 +159,62 @@ void ram_func flash_erase_blocks(uint32_t offset, size_t count)
...
+
+	flash_init_boot2_copyout();
+
+	__compiler_memory_barrier();
+
+	flash_connect_internal_func();
+	flash_exit_xip_func();
+
+	flash_cs_force(0);
+    size_t tx_remaining = count;
```
Comment (deadprogram) [inline src/machine/machine_rp2040_rom.go]: Formatting?

Comment (kenbell) [reply src/machine/machine_rp2040_rom.go]: ack

File: src/machine/machine_rp2040_flash.go (hunk 2 above)
Comment (kenbell) [reply src/machine/machine_rp2040_flash.go]: i was less sure about these - i'll think about it.  It's a bit messier given the flash 'protocol stuff' and returning an unaligned slice into it.  I'm wondering if escape analysis would take care of allocating rxBuf on the stack?

Comment (kenbell) [reply src/machine/machine_rp2040_flash.go]: Looks like we can use just one buffer for tx and rx and return a slice of that, so I've done that.

**Comment by deadprogram** [comment]
Hopefully last feedback: should we perhaps call this something less generic than `HardwareID`? Possible options that come to mind:

- `SerialNumber`
- `UniqueID`
- `DeviceID`

**Comment by kenbell** [comment]
I like DeviceID - so I’ve gone with that

**Comment by deadprogram** [comment]
Thanks very much for this addition @kenbell now merging.

**Pull request merged**